File: CS/Form1.cs
using DevExpress.DataAccess.ConnectionParameters;
using DevExpress.DataAccess.Sql;
using DevExpress.XtraReports.UI;
using System;
using System.Collections.Generic;
using System.IO;

namespace XRTable_RuntimeCreation
{
    public partial class Form1 : DevExpress.XtraEditors.XtraForm
    {
        public Form1() {
            InitializeComponent();
        }

        private void btnTableInCode_Click(object sender, EventArgs e) {
            XtraReport report = CreateEmptyReport();
            
            AddTablesToReport(report);

            ReportPrintTool printTool = new ReportPrintTool(report);
            printTool.ShowPreviewDialog();
        }
        private void btnTableInEventHandler_Click(object sender, EventArgs e) {
            XtraReport report = CreateEmptyReport();
            report.BeforePrint += Report_BeforePrint;

            ReportPrintTool printTool = new ReportPrintTool(report);
            printTool.ShowPreviewDialog();
        }

        private void btnLoadReportCodeBehind_Click(object sender, EventArgs e)
        {
            TableReport report = new TableReport();
            report.ShowPreviewDialog();
        }
        public static void AddTablesToReport(XtraReport report) {
            List<string> fields = new List<string>() {
                "CompanyName",
                "ContactName",
                "Phone",
                "City"
            };

            XRTable headerTable = GetHeaderTable(fields, 
                report.PageWidth - report.Margins.Left - report.Margins.Right);
            XRTable table = GetTableBoundToData(fields, 
                report.PageWidth - report.Margins.Left - report.Margins.Right);

            DetailBand detailBand = report.Bands.GetBandByType(typeof(DetailBand)) as DetailBand;
            PageHeaderBand pageHeaderBand = 
                report.Bands.GetBandByType(typeof(PageHeaderBand)) as PageHeaderBand;

            pageHeaderBand.Controls.Add(headerTable);
            detailBand.Controls.Add(table);

            pageHeaderBand.HeightF = headerTable.HeightF;
            detailBand.HeightF = table.HeightF;
        }
        public static XRTable GetHeaderTable(List<string> fields, float tableSize) {
            var table = new XRTable();
            
            table.BeginInit();
            
            table.LocationF = new DevExpress.Utils.PointFloat(0F, 0F);
            table.Borders = DevExpress.XtraPrinting.BorderSide.All;
            
            var tableRow = new XRTableRow();
            float cellSize = tableSize / fields.Count;

            foreach(var field in fields) {
                var cell = new XRTableCell() {
                    Text = field,
                    WidthF = cellSize,
                    BackColor = System.Drawing.Color.Gray
                };
                tableRow.Cells.Add(cell);
            }            
            
            table.Rows.Add(tableRow);

            table.AdjustSize();
            
            table.EndInit();
            
            return table;
        }
        public static XRTable GetTableBoundToData(List<string> fields, float tableSize) {
            var table = new XRTable();

            table.BeginInit();

            table.LocationF = new DevExpress.Utils.PointFloat(0F, 0F);
            table.Borders = DevExpress.XtraPrinting.BorderSide.Left 
                | DevExpress.XtraPrinting.BorderSide.Right 
                | DevExpress.XtraPrinting.BorderSide.Bottom;

            var tableRow = new XRTableRow();
            float cellSize = tableSize / fields.Count;

            foreach(var field in fields) {
                var cell = new XRTableCell() {
                    Text = field,
                    WidthF = cellSize
                };
                cell.ExpressionBindings.Add(new ExpressionBinding("Text", $"[{field}]"));
                tableRow.Cells.Add(cell);
            }

            table.Rows.Add(tableRow);

            table.AdjustSize();
            table.EndInit();

            return table;
        }
        public static XtraReport CreateEmptyReport() {
            XtraReport report = new XtraReport() {
                DataSource = CreateSqlDataSource(),
                DataMember = "Customers"
            };
            var pageHeaderBand = new PageHeaderBand();
            var detailBand = new DetailBand();
            report.Bands.AddRange(new Band[] { pageHeaderBand, detailBand });
            return report;
        }
        public static SqlDataSource CreateSqlDataSource() {
            string rootPath = Path.GetDirectoryName(typeof(Form1).Assembly.Location);
            
            SQLiteConnectionParameters connectionParameters = 
                new SQLiteConnectionParameters(Path.Combine(rootPath, "nwind.db"), null);
            SqlDataSource sqlDataSource = new SqlDataSource(connectionParameters);

            SelectQuery querySalesPerson = SelectQueryFluentBuilder
                .AddTable("Customers")
                .SelectAllColumnsFromTable()
                .Build("Customers");
            sqlDataSource.Queries.Add(querySalesPerson);
            return sqlDataSource;
        }
        private void Report_BeforePrint(object sender, System.ComponentModel.CancelEventArgs e) {
            var report = sender as XtraReport;
            AddTablesToReport(report);
        }

    }
}

File: VB/Form1.vb
Imports DevExpress.DataAccess.ConnectionParameters
Imports DevExpress.DataAccess.Sql
Imports DevExpress.XtraReports.UI
Imports System
Imports System.Collections.Generic
Imports System.IO

Namespace XRTable_RuntimeCreation
	Partial Public Class Form1
		Inherits DevExpress.XtraEditors.XtraForm

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub btnTableInCode_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnTableInCode.Click
			Dim report As XtraReport = CreateEmptyReport()

			AddTablesToReport(report)

			Dim printTool As New ReportPrintTool(report)
			printTool.ShowPreviewDialog()
		End Sub
		Private Sub btnTableInEventHandler_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnTableInEventHandler.Click
			Dim report As XtraReport = CreateEmptyReport()
			AddHandler report.BeforePrint, AddressOf Report_BeforePrint

			Dim printTool As New ReportPrintTool(report)
			printTool.ShowPreviewDialog()
		End Sub

		Private Sub btnLoadReportCodeBehind_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnLoadReportCodeBehind.Click
			Dim report As New TableReport()
			report.ShowPreviewDialog()
		End Sub
		Public Shared Sub AddTablesToReport(ByVal report As XtraReport)
			Dim fields As New List(Of String)() From {"CompanyName", "ContactName", "Phone", "City"}

			Dim headerTable As XRTable = GetHeaderTable(fields, report.PageWidth - report.Margins.Left - report.Margins.Right)
			Dim table As XRTable = GetTableBoundToData(fields, report.PageWidth - report.Margins.Left - report.Margins.Right)

			Dim detailBand As DetailBand = TryCast(report.Bands.GetBandByType(GetType(DetailBand)), DetailBand)
			Dim pageHeaderBand As PageHeaderBand = TryCast(report.Bands.GetBandByType(GetType(PageHeaderBand)), PageHeaderBand)

			pageHeaderBand.Controls.Add(headerTable)
			detailBand.Controls.Add(table)

			pageHeaderBand.HeightF = headerTable.HeightF
			detailBand.HeightF = table.HeightF
		End Sub
		Public Shared Function GetHeaderTable(ByVal fields As List(Of String), ByVal tableSize As Single) As XRTable
			Dim table = New XRTable()

			table.BeginInit()

			table.LocationF = New DevExpress.Utils.PointFloat(0F, 0F)
			table.Borders = DevExpress.XtraPrinting.BorderSide.All

			Dim tableRow = New XRTableRow()
			Dim cellSize As Single = tableSize / fields.Count

			For Each field In fields
				Dim cell = New XRTableCell() With {.Text = field, .WidthF = cellSize, .BackColor = System.Drawing.Color.Gray}
				tableRow.Cells.Add(cell)
			Next field

			table.Rows.Add(tableRow)

			table.AdjustSize()

			table.EndInit()

			Return table
		End Function
		Public Shared Function GetTableBoundToData(ByVal fields As List(Of String), ByVal tableSize As Single) As XRTable
			Dim table = New XRTable()

			table.BeginInit()

			table.LocationF = New DevExpress.Utils.PointFloat(0F, 0F)
			table.Borders = DevExpress.XtraPrinting.BorderSide.Left Or DevExpress.XtraPrinting.BorderSide.Right Or DevExpress.XtraPrinting.BorderSide.Bottom

			Dim tableRow = New XRTableRow()
			Dim cellSize As Single = tableSize / fields.Count

			For Each field In fields
				Dim cell = New XRTableCell() With {.Text = field, .WidthF = cellSize}
				cell.ExpressionBindings.Add(New ExpressionBinding("Text", $"[{field}]"))
				tableRow.Cells.Add(cell)
			Next field

			table.Rows.Add(tableRow)

			table.AdjustSize()
			table.EndInit()

			Return table
		End Function
		Public Shared Function CreateEmptyReport() As XtraReport
			Dim report As New XtraReport() With {.DataSource = CreateSqlDataSource(), .DataMember = "Customers"}
			Dim pageHeaderBand = New PageHeaderBand()
			Dim detailBand = New DetailBand()
			report.Bands.AddRange(New Band() { pageHeaderBand, detailBand })
			Return report
		End Function
		Public Shared Function CreateSqlDataSource() As SqlDataSource
			Dim rootPath As String = Path.GetDirectoryName(GetType(Form1).Assembly.Location)

			Dim connectionParameters As New SQLiteConnectionParameters(Path.Combine(rootPath, "nwind.db"), Nothing)
			Dim sqlDataSource As New SqlDataSource(connectionParameters)

			Dim querySalesPerson As SelectQuery = SelectQueryFluentBuilder.AddTable("Customers").SelectAllColumnsFromTable().Build("Customers")
			sqlDataSource.Queries.Add(querySalesPerson)
			Return sqlDataSource
		End Function
		Private Sub Report_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs)
			Dim report = TryCast(sender, XtraReport)
			AddTablesToReport(report)
		End Sub

	End Class
End Namespace
